Hoisting common expressions out of If blocks

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #59492

Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29399533](https://our.internmc.facebook.com/intern/diff/D29399533)

Gamrix committed:
Hoisting common expressions out of If blocks

Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -529,13 +434,16 @@ void ProfilingGraphExecutorImpl::runProfilingInsensitiveOptimizations(
   EliminateCommonSubexpression(graph);
-  GRAPH_DEBUG(
-      "After EliminateCommonSubexpression, before ConstantPooling\n", *graph);
+  GRAPH_DEBUG("After HoistCommonExpression, before ConstantPooling\n", *graph);
```
Note to self: small nit here

Gamrix commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -268,112 +269,16 @@ void runPreAutodiffPassPipeline(std::shared_ptr<Graph>& graph) {
       "After CheckInplace (end of runPreAutodiffPassPipeline)\n", *graph);
 }
 
-void runDiffGraphPasses(std::shared_ptr<Graph>& graph) {
```
Unused function

Krovatkin replied:
wtf. Errr... I'll check the history tomorrow

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on test/jit/test_if_hoisting.py:
```diff
@@ -0,0 +1,120 @@
+        self.assertEqual(fn(True, 1), 4)
+
+    def test_if_hoist_transposed_expr(self):
+        # Making sure that the if statement
```
what are we trying to test here ?

Gamrix replied:
oops, the comment is wrong, We are trying to test that the you can still eliminate a line of code if the statement is not already at the front of one of the if blocks.

eellison commented on test/jit/test_if_hoisting.py:
```diff
@@ -0,0 +1,120 @@
+        FileCheck().check_count("prim::If", 0, exactly=True).run(op_graph)
+        FileCheck().check_count("aten::add", 1, exactly=True).run(op_graph)
+
+        self.assertEqual(fn(True, 1), (4, 2))
```
nit: use `checkScript` or something like, `self.assertEqual(fn_script(True, 1), fn)` instead of manually specifying output (one less thing to think about...)

eellison commented on test/jit/test_if_hoisting.py:
```diff
@@ -0,0 +1,120 @@
+        self.assertEqual(fn(False, 5), (10, 8))
+
+    def test_if_hoist_shadowed_expr(self):
+        # Making sure that multiple forms of shadowing are handled correctly
```
what does shadowing mean here ?

Gamrix replied:
I meant that the variable is reused. Shadowing is likely the wrong term for that.

eellison reviewed (commented):
Looks great!! 

A couple other comments:
Do you mind using the `getOrCreateAliasDb` idiom, similar to some of the other passes ? Creating alias db is expensive, and we expect to rarely find expressions that we hoist out of if nodes. Ideally we do not always instantiate an alias db. 

Also, with the getOrCreateAliasDb change, you might find it easier to write this in pass in a class, as with the other passes.

Gamrix commented on torch/csrc/jit/ir/node_hashing.cpp:
```diff
@@ -203,6 +203,10 @@ bool attributesEqualCSE(const Node* lhs, const Node* rhs) {
 } // anonymous namespace
 
+// Makes a hash where hashes are equal when they can be
+// merged into the same expression. Notably, the output variable
```
The output variable only needs to be the same type, not the same variable.

Gamrix commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,111 @@
+
+    // fix up the if block outputs
+
+    true_block_nodes.erase(true_b_node);
```
I don't understand why that is necessary.

eellison replied:
we initialize `true_block_nodes` at the start of the function, so i think anytime we move a node out of the if node and into it's surrounding block we are invalidating the hash of all nodes in `true_block_nodes` which have as input any outputs of the newly moved node

Krovatkin replied:
this sounds important. It sounds like we will fail to hoist out any expressions whose inputs have already been hoisted out.

Gamrix replied:
Yeah, didn't think about how the hash would change. 

Is it better to remove the items, and then readd them when we change the inputs, or to just remove the set completely and just loop over the if block (of course the latter has a higher big-O complexity)

Gamrix replied:
(also was able to repro this failure with an additional test :) )

eellison replied:
Yes, the complexity with removing and re adding is O(#num changes) * O(# avg number of node users in block) (typically 1-2) instead of O(#num changes) * O(# of nodes in block)

Gamrix commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,111 @@
+    // Need to check both sides match to eliminate common if block outputs
+    Value* true_block_output = true_block->outputs().at(i);
+    Value* false_block_output = false_block->outputs().at(i);
+    if (true_block_output != false_block_output ||
```
So what about the example:

```
if x:
  a = b + c
  a.add_inplace(d)
else:
  a = b + c
return a
```

eellison replied:
What we are doing here is just cleaning up the IR, we are not affecting the way anything is computed. that is, at this point in the pass, we are just updating the IR to reflect that `a` is computed and in scope before the if node. this point is maybe a little subtle, we can talk about it more maybe in person.

however,  this should be fine too
-> 
```
a = b + c
if x:
  a.add_inplace(d)
return a
```
has same exact semantics

see other comment about hasWriters potentially not being needed as a check.

Gamrix commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,111 @@
+    if (node->kind() != prim::If) {
+      continue;
+    }
+    // Note that they will not have the same outputs
```
That comment should just be deleted.

Krovatkin commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,111 @@
+    true_block->eraseOutput(i);
+    false_block->eraseOutput(i);
+  }
+  // We will let DCE elimiate the if block if it should be eliminated.
```
> // We will let DCE elimiate the if block if it should be eliminated.

if it's empty, maybe? otherwise it sounds a bit circular.

Gamrix commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,111 @@
+      continue;
+    }
+    // We have a matching output, and can hoist out of if block
+    if_node->outputs().at(i)->replaceAllUsesWith(true_block_output);
```
@eellison  I am still not convinced that I have done enough checks on the `IF` output to make sure that this change is safe. Can you confirm that there are no more checks I need to do on the `if_node->outputs().at(i)` value?

eellison replied:
will look into it when i do full re-review. just request when it's ready;

![image](https://user-images.githubusercontent.com/11477974/121587783-05b1ca00-c9ea-11eb-8755-548524205b22.png)

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on torch/csrc/jit/ir/node_hashing.cpp:
```diff
@@ -230,6 +232,8 @@ size_t HashNode::operator()(const Node* k) const {
       constant_hash);
 };
 
+// Checks that two nodes have the same imputs, output types
```
nit: `imputs` -> `inputs`

eellison commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,142 @@
+  CommonExpressionHoister(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
```
nit: you might find it easier to just keep `changed_` as a struct member..

eellison commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,142 @@
+  // if block out of the if block.
+
+  GRAPH_DUMP("Before CSE", graph);
+  CommonExpressionHoister ceh(graph);
```
nit: maybe add a post dump as well..

eellison commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,142 @@
+      std::unordered_set<Node*> true_b_uses;
+      for (Value* true_out : true_b_node->outputs()) {
+        for (Use true_use : true_out->uses()) {
+          true_b_uses.insert(true_use.user);
```
we also want to check that the uses are on the right block, you can check `owningBlock()`

Gamrix replied:
Why would the use of the variable not be in that block or a compatible outer block?

eellison replied:
```
prim::If
    block1():
          a = op()
          op1(a)
          prim::If
                 block2():
                        op2(a)
```

We don't want to add recursively contained uses, e.g. we just want uses on block1, not block2. The nodes in block2 are not meant to be in the true_block_nodes

eellison commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    // fix up the if block outputs
+    for (size_t i = 0; i < true_block->outputs().size();) {
```
nit: maybe refactor this to its own function

eellison commented on torch/csrc/jit/runtime/graph_executor.cpp:
```diff
@@ -919,7 +920,10 @@ void runOptimization(
-      "After EliminateCommonSubexpression, before PeepholeOptimize\n", *graph);
+      "After EliminateCommonSubexpression, before HoistCommonExpression\n",
+      *graph);
+  HoistCommonExpression(graph);
```
i think just running this once in the pass is fine. this is a very targeted optimization to try to handle the case of 
```
if inplace:
    y = x.add_(3)
else:
    y = x.add(3)
```

which we want to rewrite to 
```
if inplace:
    y = x.add(3)
else:
    y = x.add(3)
```
and hoist out. we're not running the mutation removal yet, but we should be shortly... cc @desertfire make sure these line up with the mutation removal pass. 

we're not really expecting any juice out of this pass other than the inplace handling so just once fine probably

eellison reviewed (approved):
Ship it 🚀🚀🚀 There is just the `owningBlock` check we need to add. 

In case you're unfamiliar, the general protocal of pytorch reviews is you accept if you're reasonably confident that the reviewer can handle the remaining changes without another look at the code, and you're fine with the nits potentially being addressed or not.

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29399533).

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29399533](https://our.internmc.facebook.com/intern/diff/D29399533)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29399533).

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29399533](https://our.internmc.facebook.com/intern/diff/D29399533)

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29399533](https://our.internmc.facebook.com/intern/diff/D29399533)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29399533).

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29399533](https://our.internmc.facebook.com/intern/diff/D29399533)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29399533).

Gamrix committed:
Update on "Hoisting common expressions out of If blocks"


Summary:
Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29399533](https://our.internmc.facebook.com/intern/diff/D29399533)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29399533).
